bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeHandler.java

```diff
@@ -0,0 +1,546 @@
+import org.eclipse.smarthome.core.library.types.QuantityType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.library.unit.ImperialUnits;
+import org.eclipse.smarthome.core.library.unit.SIUnits;
+import org.eclipse.smarthome.core.thing.Channel;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.openhab.binding.gree.internal.GreeConfiguration;
+import org.openhab.binding.gree.internal.GreeException;
+import org.openhab.binding.gree.internal.GreeTranslationProvider;
+import org.openhab.binding.gree.internal.discovery.GreeDeviceFinder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link GreeHandler} is responsible for handling commands, which are sent to one of the channels.
+ *
+ * @author John Cunha - Initial contribution
+ * @author Markus Michels - Refactoring, adapted to OH 2.5x
+ */
+@NonNullByDefault
+public class GreeHandler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(GreeHandler.class);
+    private final GreeTranslationProvider messages;
+    private final GreeDeviceFinder deviceFinder;
+    private final String thingId;
+    private GreeConfiguration config = new GreeConfiguration();
+    private GreeAirDevice device = new GreeAirDevice();
+    private Optional<DatagramSocket> clientSocket = Optional.empty();
+    private boolean forceRefresh = false;
+
+    private @Nullable ScheduledFuture<?> refreshTask;
+    private long lastRefreshTime = 0;
+
+    public GreeHandler(Thing thing, GreeTranslationProvider messages, GreeDeviceFinder deviceFinder) {
+        super(thing);
+        this.messages = messages;
+        this.deviceFinder = deviceFinder;
+        this.thingId = getThing().getUID().getId();
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(GreeConfiguration.class);
+        if (config.ipAddress.isEmpty() || (config.refresh < 0)) {
+            String message = messages.get("thinginit.invconf");
+            logger.warn("{}: {}", thingId, message);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, message);
+        }
+
+        // set the thing status to UNKNOWN temporarily and let the background task decide for the real status.
+        // the framework is then able to reuse the resources from the thing handler initialization.
+        updateStatus(ThingStatus.UNKNOWN);
+
+        // Start the automatic refresh cycles
+        startAutomaticRefresh();
+        scheduler.execute(this::initializeThing);
```

Comment:
You should make sure that this returns a future so you can cancel it on dispose.

Reply:
what do you mean?
- startAutomaticRefresh() creates the refreshTask
- scheduler.execute() does not return a ScheduledFuture

that's what I also saw in other bindings

Reply:
```suggestion
        initializeFuture = scheduler.submit(this::initializeThing);
```

Reply:
changed